clients/src/main/java/org/apache/kafka/clients/consumer/internals/FetchBuffer.java

```diff
@@ -166,7 +169,7 @@ void awaitNotEmpty(Timer timer) {
         try {
             lock.lock();
 
-            while (isEmpty()) {
+            while (isEmpty() && !wakedUp.compareAndSet(true, false)) {
```

Comment:
Nice! I was first convinced that we have an interleaving here that misses the wakeup, but `wakedUp` takes care.